docs/custom.md

```diff
@@ -295,49 +295,42 @@ invoked by the IBC module after the packet has been proved valid and correctly p
 keepers. Thus, the `OnRecvPacket` callback only needs to worry about making the appropriate state
 changes given the packet data without worrying about whether the packet is valid or not.
 
-Modules may return an acknowledgement as a byte string and return it to the IBC handler.
+Modules may return to the IBC handler an acknowledgement which implements the Acknowledgement interface.
 The IBC handler will then commit this acknowledgement of the packet so that a relayer may relay the
 acknowledgement back to the sender module.
 
+The state changes that occurred during this callback will only be written if:
+- the acknowledgement was successful as indicated by the `Success()` function of the acknowledgement
+- if the acknowledgement returned is nil indicating that an asynchronous process is occurring
+
 ```go
 OnRecvPacket(
     ctx sdk.Context,
     packet channeltypes.Packet,
-) (res *sdk.Result, ack []byte, abort error) {
+) ibcexported.Acknowledgement {
     // Decode the packet data
     packetData := DecodePacketData(packet.Data)
 
-    // do application state changes based on packet data
-    // and return result, acknowledgement and abortErr
-    // Note: abortErr is only not nil if we need to abort the entire receive packet, and allow a replay of the receive.
-    // If the application state change failed but we do not want to replay the packet,
-    // simply encode this failure with relevant information in ack and return nil error
-    res, ack, abortErr := processPacket(ctx, packet, packetData)
-
-    // if we need to abort the entire receive packet, return error
-    if abortErr != nil {
-        return nil, nil, abortErr
-    }
+    // do application state changes based on packet data and return the acknowledgement
+    // NOTE: The acknowledgement will indicate to the IBC handler if the application 
+    // state changes should be written via the `Success()` function. Application state
+    // changes are only written if the acknowledgement is successful or the acknowledgement
+    // returned is nil indicating that an asynchronous acknowledgement will occur.
+    ack := processPacket(ctx, packet, packetData)
```

Comment:
```suggestion
    ack := processPacket(ctx, packet)
```

It's a bit confusing to have both `packet` and `packetData` here. For documentation purposes, we can do keep it as above. `processPacket` is responsible for decoding packetdata

Reply:
I think it is useful to show the app developer they need to decode the packet data, as shown by the line above this one

Going to leave as is since it matches ICS 20